paasta_tools/vitesscluster_tools.py

```diff
@@ -15,6 +17,594 @@
+    "keep_logs": "72h",
+    "enable-lag-throttler": "true",
+    "throttle_check_as_check_self": "true",
+    "throttle_metrics_query": "",
+    "throttle_metrics_threshold": "",
+    "db_charset": "utf8mb4",
+    "disable_active_reparents": "true",
+}
+
+
+def build_affinity_spec(paasta_pool):
+    spec = {
+        "nodeAffinity": {
+            "requiredDuringSchedulingIgnoredDuringExecution": {
+                "nodeSelectorTerms": [
+                    {
+                        "matchExpressions": [
+                            {
+                                "key": "yelp.com/pool",
+                                "operator": "In",
+                                "values": [paasta_pool],
+                            }
+                        ]
+                    }
+                ]
+            }
+        }
+    }
+    return spec
+
+
+def build_extra_labels(paasta_pool, paasta_cluster):
+    """
+    Build extra labels to adhere to paasta contract
+    """
+    extra_labels = {
+        "yelp.com/owner": "dre_mysql_working_hours",
+        "paasta.yelp.com/cluster": paasta_cluster,
+        "paasta.yelp.com/pool": paasta_pool,
+    }
+    return extra_labels
+
+
+def build_extra_env(paasta_cluster):
+    """
+    Build extra env to adhere to paasta contract
+    """
+    extra_env = [
+        {
+            "name": "PAASTA_POD_IP",
+            "valueFrom": {"fieldRef": {"fieldPath": "status.podIP"}},
+        },
+        {"name": "PAASTA_CLUSTER", "value": paasta_cluster},
+    ]
+    return extra_env
+
+
+def build_cell_config(cell, paasta_pool, paasta_cluster, region):
+    """
+    Build vtgate config
+    """
+    config = {
+        "name": cell,
+        "gateway": {
+            "extraFlags": {
+                "mysql_auth_server_impl": "vault",
+                "mysql_auth_vault_addr": f"https://vault-dre.{region}.yelpcorp.com:8200",
+                "mysql_auth_vault_path": "secrets/vitess/vt-gate/vttablet_credentials.json",
+                "mysql_auth_vault_tls_ca": f"/etc/vault/all_cas/acm-privateca-{region}.crt",
+                "mysql_auth_vault_ttl": "60s",
+            },
+            "affinity": build_affinity_spec(paasta_pool),
+            "extraLabels": build_extra_labels(paasta_pool, paasta_cluster),
+            "extraEnv": build_extra_env(paasta_cluster),
+            "replicas": 1,
+            "resources": {
+                "requests": {
+                    "cpu": "100m",
+                    "memory": "256Mi",
+                },
+                "limits": {"memory": "256Mi"},
+            },
+        },
+    }
+    vtgate_extra_env = copy.deepcopy(VTGATE_EXTRA_ENV)
+    for env in vtgate_extra_env:
+        if env["name"] == "VAULT_ADDR":
+            env["value"] = f"https://vault-dre.{region}.yelpcorp.com:8200"
+        if env["name"] == "VAULT_CACERT":
+            env["value"] = f"/etc/vault/all_cas/acm-privateca-{region}.crt"
+        config["gateway"]["extraEnv"].append(env)
```

Comment:
(and especially since we have to do this for loop in a couple other functions)

Reply:
These default extra env are still a dict, but yes, makes sense to have a common function to update or produce a new dict with overrides since it's done in building spec for multiple components differently, vtgate, vttablet, vtadmin etc.,

Reply:
https://github.com/Yelp/paasta/pull/3889/commits/0f12a638f8b31b69697a4d9e5ae42843a5adda32